playground/dockerfile/Dockerfile.AppHost/aspire-manifest.json

```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "https://json.schemastore.org/aspire-8.0.json",
+  "resources": {
+    "goversion": {
+      "type": "parameter.v0",
+      "value": "{goversion.inputs.value}",
+      "inputs": {
+        "value": {
+          "type": "string"
+        }
+      }
+    },
+    "mycontainer": {
+      "type": "container.v0",
+      "image": "myimage:latest",
+      "build": {
```

Comment:
@vhvb1989 @ellismg I wanted to point out some upcoming schema changes in support for allowing us to build container images. We are looking at deprecating `dockerfile.v0` in favor of unifying it with `container.v0` and adding extra properties to `container.v0` to provide enough information for deployment tools to perform a build.

This is off the back of some work that @danegsta and @karolz-ms have been doing around adding support for Docker builds in DCP. This PR is getting this working on the app model side of things.

We are targeting this for 8.1. Wanted to get your early feedback here. If we can agree on the schema I'll update the official schema.

We need to decide whether we want to publish a 8.1 schema which adds this or just add to the 8.0 schema.

Reply:
@mitchdenny What does it mean in this case since there is an `image` property on the container in addition to the `build` block?  I would have expected one or the other?  Is this trying to influence the naming of the image that is going to be pushed to the remote ACR after we build the dockerfile from the `build` section?

I think this format would work fine for us - FWIW - we already support doing a `docker build` for `dockerfile.v0` resources.  We'd treat this the same way, and I have a pretty good view about how our manifest loader could just rewrite this into the old form so we could get privates quickly for you to test.

Reply:
@ellismg yeah I'm thinking that the image name is just there to influence the name of the image that is pushed. The other aspect of this is the image version label. What do you do with dockerfile.v0 labels today?

Reply:
@mitchdenny We have been using the current UTC Unix in the generation of the name and label, so we got a unique container image per `deploy`, instead of having a fixed image name that we overwrote each time.  I'm not sure how I should think about pushing to same image over and over again.  A possible nice thing about using a unique image each time is that in something like ACA it does mean the previous revisions point to the older docker container images, which could be useful in some rollback scenario or later on if we want to start exploring something like blue/green deployments or other stagged rollout type operations?

Reply:
One option would be to apply and push more than one tag for the image. A unique deployment tag gets generated using the current pattern, but if `image` is specified, also apply and push that tag.

Reply:
> A unique deployment tag gets generated using the current pattern, but if image is specified, also apply and push that tag.

That makes sense to me - How does a user in code reflect this intention?  I assume they won't be able to do something like:

```csharp
builder.AddContainer()
       .FromDockerfile("qots")
       .WithBuildArg("GO_VERSION", goVersion);
```

(i.e. not passing any arguments to `AddContainer`, because it's not obvious from the API that if you call the "simpler" overload, you need to get on a `.FromDockerfile()` plan.

Would you imagine something like:

```csharp
builder.AddContainer("doesnt", "matter")
       .FromDockerfile("qots")
       .WithBuildArg("GO_VERSION", goVersion)
       .WithoutDefaultTag()
```
?

Reply:
What updating AddContainer to make `image` optional and add a new optional `dockerfile` field as well?

`public static IResourceBuilder<ContainerResource> AddContainer(this IDistributedApplicationBuilder builder, string name, string? image = default, string tag = "latest", string? dockerfile = default)`

```
builder.AddContainer(dockerfile: "quots")
    .WithBuildArg("GO_VERSION", goVersion);
```

It could apply `ContainerImageAnnotation`, `DockerfileBuildAnnotation`, or both depending on the arguments. And if both image, and dockerfile are all null/empty, throw an exception.

`IsContainer` and `GetContainerResources` would both need to be updated to return resources that have either `ContainerImageAnnotation` or `DockerfileBuildAnnotation`.

Reply:
It'd mean `AddContainer` always returns a "valid" container resource, but it might have just a `ContainerImageAnnotation`, just a `DockerfileBuildAnnotation`, or both. In either case `WithImageTag` or `FromDockerfile` could still be appended to modify the annotations

Reply:
Hrm ... I think I'm more comfortable with the way it is now. The way I see it is that `AddContainer(...)` is all about telling Aspire that there is a container resource in the model ... for that you need 3 essential bits of information ... the resource name, the image, and optionally the tag (with `latest`) being the default.

The purpose of `FromDockerfile` is all about telling DCP and manifest publishing how that image is sourced. By default we just try to pull it, but if we provide some additional instructions (the new annotation) the deployment tools and DCP will use to build the image locally.

When the container image is built via DCP, we probably want the tag to be unique per run. Here is why ... when we run locally if we had two AppHost projects open with the same name which happened to have two containers with the same name (not unreasonable) we wouldn't want the build/run loop to interfere with one another (so they both couldn't use latest because you might get some weird results).

For deployment tool based builds we'd probably let the tool decide and be consistent with whatever they use for project resource containers.

Reply:
It feels a little weird to force users specify an image (and potentially a tag) if we aren't going to honor it, but I definitely agree that applying a random tag per run makes sense to avoid conflicts. I think I'd expect to be able to find an image with the specified tag after an F5. I'm wondering if it makes sense for me to add a `tags` property to DCP to support more than one tag during build?

Reply:
I think we honour everything (right now) - which is fine except when the image names conflict and you get two versions of "myfavoritecontainerthatiuseinallmyprojects" both tagged latest and you can a random one loaded as part of a race condition.

Another alternative might be that we always apply the tag as specified, but when the container is launched we use the SHA that DCP would know from doing the build. That would avoid any conflicts.

No need to invent another mechanism to uniquely identify an image from the build. Seem fair?

Reply:
Using a sha is potentially a good idea for deployment tools too.